### ISSUE TITLE
Unable to start a WebDriver session using directConnect

### ISSUE THREAD

**Issue by JoshSchreuder** [angular/protractor#1731]
I just upgraded from v1.0.0 to v1.6.0, all went smoothly and none of my tests broke. I was keen to try out the new `directConnect` method which looks more stable than running Selenium server.

Anyway, my tests run on my CI server, and this worked once, the very first time I tried with directConnect, however all subsequent runs have produced:

```
[20:46:42]Starting: C:\Windows\system32\cmd.exe /c grunt e2e:frontend --url=http://mysite.com --stack
[20:46:42]in directory: D:\TeamCity\buildAgent\work\490b16ab3134ee36\Build
[20:47:51] [4mRunning "e2e:frontend" (e2e) task [24m
[20:47:53]
[20:47:53] [4mRunning "protractor:frontend" (protractor) task [24m
[20:48:04][launcher] Running 2 instances of WebDriver
[20:50:30]
[20:50:30]------------------------------------
[20:50:30][chrome #1] PID: 4976
[20:50:30][chrome #1] Using ChromeDriver directly...
[20:50:30][chrome #1] ERROR - Unable to start a WebDriver session.
[20:50:30]
[20:50:30][launcher] Runner process exited unexpectedly with error code: 1
[20:50:30][launcher] 1 instance(s) of WebDriver still running
[20:53:46] [31m>>  [39m
[20:53:46] [31mFatal error: protractor exited with code: 100  [39m
```

I'm not sure why this is happening - the environment has not been modified since I ran the tests successfully the first time.

Is there a way I can get more detailed logs on this issue to help diagnose?

**Comment by juliemr** [comment]
Can you try adding the `--troubleshoot` flag to your call to protractor, to get some more logging info?

**Comment by JoshSchreuder** [comment]
Sure, here is the result:
```
[10:23:27]Starting: C:\Windows\system32\cmd.exe /c grunt e2e:frontend --url=http://mysite.com --troubleshoot
[10:23:27]in directory: D:\TeamCity\buildAgent\work\490b16ab3134ee36\Build
[10:23:48] [4mRunning "e2e:frontend" (e2e) task [24m
[10:23:48]
[10:23:48] [4mRunning "protractor:frontend" (protractor) task [24m
[10:23:49]DEBUG - Running with --troubleshoot
[10:23:49]DEBUG - Protractor version: 1.6.0
[10:23:49]DEBUG - Your base url for tests is http://mysite.com
[10:23:49][launcher] Running 2 instances of WebDriver
[10:24:34]
[10:24:34]------------------------------------
[10:24:34][chrome #1] PID: 3088
[10:24:34][chrome #1] Using ChromeDriver directly...
[10:24:34][chrome #1] DEBUG - WebDriver session successfully started with capabilities { caps_: 
[10:24:34][chrome #1]    { acceptSslCerts: true,
[10:24:34][chrome #1]      applicationCacheEnabled: false,
[10:24:34][chrome #1]      browserConnectionEnabled: false,
[10:24:34][chrome #1]      browserName: 'chrome',
[10:24:34][chrome #1]      chrome: { userDataDir: 'D:\\TeamCity\\buildAgent\\temp\\buildTmp\\scoped_dir3572_21545' },
[10:24:34][chrome #1]      cssSelectorsEnabled: true,
[10:24:34][chrome #1]      databaseEnabled: false,
[10:24:34][chrome #1]      handlesAlerts: true,
[10:24:34][chrome #1]      javascriptEnabled: true,
[10:24:34][chrome #1]      locationContextEnabled: true,
[10:24:34][chrome #1]      mobileEmulationEnabled: false,
[10:24:34][chrome #1]      nativeEvents: true,
[10:24:34][chrome #1]      platform: 'Windows NT',
[10:24:34][chrome #1]      rotatable: false,
[10:24:34][chrome #1]      takesHeapSnapshot: true,
[10:24:34][chrome #1]      takesScreenshot: true,
[10:24:34][chrome #1]      version: '39.0.2171.99',
[10:24:34][chrome #1]      webStorageEnabled: true } }
[10:24:34][chrome #1] 
[10:24:34]Running Jasmine Tests
[10:24:34]
[10:24:34][launcher] Runner process exited unexpectedly with error code: 1
[10:24:34][launcher] 1 instance(s) of WebDriver still running
[10:28:03]........................... [31m>>  [39m
[10:28:03] [31mFatal error: protractor exited with code: 100  [39m
```

**Comment by JoshSchreuder** [comment]
Of course, when I run it manually it works just fine :disappointed: 
```
D:\TeamCity\buildAgent\work\490b16ab3134ee36\Build>grunt e2e:frontend --url=http://mysite.com --troubleshoot
Running "e2e:frontend" (e2e) task

Running "protractor:frontend" (protractor) task
DEBUG - Running with --troubleshoot
DEBUG - Protractor version: 1.6.0
DEBUG - Your base url for tests is http://mysite.com
[launcher] Running 2 instances of WebDriver
...................................................
------------------------------------
[chrome #1] PID: 4476
[chrome #1] Using ChromeDriver directly...
[chrome #1] DEBUG - WebDriver session successfully started with capabilities { caps_:
[chrome #1]    { acceptSslCerts: true,
[chrome #1]      applicationCacheEnabled: false,
[chrome #1]      browserConnectionEnabled: false,
[chrome #1]      browserName: 'chrome',
[chrome #1]      chrome: { userDataDir: 'C:\\Users\\TeamCity\\AppData\\Local\\T
mp\\2\\scoped_dir2144_2051' },
[chrome #1]      cssSelectorsEnabled: true,
[chrome #1]      databaseEnabled: false,
[chrome #1]      handlesAlerts: true,
[chrome #1]      javascriptEnabled: true,
[chrome #1]      locationContextEnabled: true,
[chrome #1]      mobileEmulationEnabled: false,
[chrome #1]      nativeEvents: true,
[chrome #1]      platform: 'Windows NT',
[chrome #1]      rotatable: false,
[chrome #1]      takesHeapSnapshot: true,
[chrome #1]      takesScreenshot: true,
[chrome #1]      version: '39.0.2171.99',
[chrome #1]      webStorageEnabled: true } }
[chrome #1] ...........................
[chrome #1]
[chrome #1] Finished in 343.658 seconds
[chrome #1] 27 tests, 46 assertions, 0 failures
[chrome #1]

[launcher] 1 instance(s) of WebDriver still running
...
------------------------------------
[firefox #2] PID: 3288
[firefox #2] Using FirefoxDriver directly...
[firefox #2] DEBUG - WebDriver session successfully started with capabilities {
caps_:
[firefox #2]    { cssSelectorsEnabled: true,
[firefox #2]      browserName: 'firefox',
[firefox #2]      handlesAlerts: true,
[firefox #2]      javascriptEnabled: true,
[firefox #2]      nativeEvents: false,
[firefox #2]      platform: 'WINDOWS',
[firefox #2]      rotatable: false,
[firefox #2]      takesScreenshot: true,
[firefox #2]      version: '34.0.5',
[firefox #2]      webStorageEnabled: true,
[firefox #2]      applicationCacheEnabled: true,
[firefox #2]      databaseEnabled: true,
[firefox #2]      locationContextEnabled: true,
[firefox #2]      acceptSslCerts: true } }
[firefox #2] ...........................
[firefox #2]
[firefox #2] Finished in 333.858 seconds
[firefox #2] 27 tests, 46 assertions, 0 failures
[firefox #2]

[launcher] 0 instance(s) of WebDriver still running
[launcher] chrome #1 passed
[launcher] firefox #2 passed

Done, without errors.
```

**Comment by juliemr** [comment]
Hmm, the capabilities where it's looking for the Chrome binary are different, but I'm not sure if that's what's causing the error.

**Comment by hankduan** [comment]
It looks like during your two failed runs (one with troubleshoot and one without), the test died at different points. 

First died around the time of protractor trying to create a session (step 2 of runner.js) and the second when jasmine began running (step 4 of runner.js). 

Can you try running this a few times and see if it dies at different points each time?

**Comment by JoshSchreuder** [comment]
@hankduan, a couple of results from overnight:

1 build ran unit tests successfully (however, only ran Chrome for some reason?)
```
[21:58:16] [4mRunning "e2e:frontend" (e2e) task [24m
[21:58:16]
[21:58:16] [4mRunning "protractor:frontend" (protractor) task [24m
[21:58:17]DEBUG - Running with --troubleshoot
[21:58:17]DEBUG - Protractor version: 1.6.0
[21:58:17]DEBUG - Your base url for tests is http://mysite.com
[21:58:17][launcher] Running 2 instances of WebDriver
[22:03:23].....................................................
[22:03:23]------------------------------------
[22:03:23][chrome #1] PID: 2524
[22:03:23][chrome #1] Using ChromeDriver directly...
[22:03:23][chrome #1] DEBUG - WebDriver session successfully started with capabilities { caps_: 
[22:03:23][chrome #1]    { acceptSslCerts: true,
[22:03:23][chrome #1]      applicationCacheEnabled: false,
[22:03:23][chrome #1]      browserConnectionEnabled: false,
[22:03:23][chrome #1]      browserName: 'chrome',
[22:03:23][chrome #1]      chrome: { userDataDir: 'D:\\TeamCity\\buildAgent\\temp\\buildTmp\\scoped_dir6020_23918' },
[22:03:23][chrome #1]      cssSelectorsEnabled: true,
[22:03:23][chrome #1]      databaseEnabled: false,
[22:03:23][chrome #1]      handlesAlerts: true,
[22:03:23][chrome #1]      javascriptEnabled: true,
[22:03:23][chrome #1]      locationContextEnabled: true,
[22:03:23][chrome #1]      mobileEmulationEnabled: false,
[22:03:23][chrome #1]      nativeEvents: true,
[22:03:23][chrome #1]      platform: 'Windows NT',
[22:03:23][chrome #1]      rotatable: false,
[22:03:23][chrome #1]      takesHeapSnapshot: true,
[22:03:23][chrome #1]      takesScreenshot: true,
[22:03:23][chrome #1]      version: '39.0.2171.99',
[22:03:23][chrome #1]      webStorageEnabled: true } }
[22:03:23][chrome #1] 
[22:03:23]Running Jasmine Tests
[22:03:23][chrome #1]  [32m. [0m [32m. [0m [32m. [0m [32m. [0m [32m. [0m
[22:03:23]dashboard application list
[22:03:23][chrome #1]  [32m. [0m [32m. [0m [32m. [0m [32m. [0m [32m. [0m [32m. [0m [32m. [0m [32m. [0m [32m. [0m [32m. [0m [32m. [0m
[22:03:23]frontend login
[22:03:23][chrome #1]  [32m. [0m [32m. [0m [32m. [0m [32m. [0m [32m. [0m [32m. [0m [32m. [0m [32m. [0m [32m. [0m [32m. [0m [32m. [0m
[22:03:23]dashboard homepage
[22:03:23][chrome #1] 
[22:03:23][chrome #1] 
[22:03:23][chrome #1] 
[22:03:23][chrome #1] Finished in 277.246 seconds
[22:03:23][chrome #1]  [32m27 tests, 46 assertions, 0 failures
[22:03:23] [0m[chrome #1] 
[22:03:23]
[22:03:23][launcher] 1 instance(s) of WebDriver still running
[22:03:25].
[22:03:25] [32mDone, without errors. [39m
```

One run failed:
```
[21:34:59] [4mRunning "e2e:frontend" (e2e) task [24m
[21:34:59]
[21:34:59] [4mRunning "protractor:frontend" (protractor) task [24m
[21:35:03]DEBUG - Running with --troubleshoot
[21:35:03]DEBUG - Protractor version: 1.6.0
[21:35:03]DEBUG - Your base url for tests is http://mysite.com
[21:35:03][launcher] Running 2 instances of WebDriver
[21:36:28]
[21:36:28]------------------------------------
[21:36:28][chrome #1] PID: 4580
[21:36:28][chrome #1] Using ChromeDriver directly...
[21:36:28]
[21:36:28][launcher] Runner process exited unexpectedly with error code: 1
[21:36:28][launcher] 1 instance(s) of WebDriver still running
[21:40:51]........................... [31m>>  [39m
[21:40:51] [31mFatal error: protractor exited with code: 100  [39m
[21:40:51]Process exited with code 3
```

And another one failed:
```
[10:23:48] [4mRunning "e2e:frontend" (e2e) task [24m
[10:23:48]
[10:23:48] [4mRunning "protractor:frontend" (protractor) task [24m
[10:23:49]DEBUG - Running with --troubleshoot
[10:23:49]DEBUG - Protractor version: 1.6.0
[10:23:49]DEBUG - Your base url for tests is undefined
[10:23:49][launcher] Running 2 instances of WebDriver
[10:24:34]
[10:24:34]------------------------------------
[10:24:34][chrome #1] PID: 3088
[10:24:34][chrome #1] Using ChromeDriver directly...
[10:24:34][chrome #1] DEBUG - WebDriver session successfully started with capabilities { caps_: 
[10:24:34][chrome #1]    { acceptSslCerts: true,
[10:24:34][chrome #1]      applicationCacheEnabled: false,
[10:24:34][chrome #1]      browserConnectionEnabled: false,
[10:24:34][chrome #1]      browserName: 'chrome',
[10:24:34][chrome #1]      chrome: { userDataDir: 'D:\\TeamCity\\buildAgent\\temp\\buildTmp\\scoped_dir3572_21545' },
[10:24:34][chrome #1]      cssSelectorsEnabled: true,
[10:24:34][chrome #1]      databaseEnabled: false,
[10:24:34][chrome #1]      handlesAlerts: true,
[10:24:34][chrome #1]      javascriptEnabled: true,
[10:24:34][chrome #1]      locationContextEnabled: true,
[10:24:34][chrome #1]      mobileEmulationEnabled: false,
[10:24:34][chrome #1]      nativeEvents: true,
[10:24:34][chrome #1]      platform: 'Windows NT',
[10:24:34][chrome #1]      rotatable: false,
[10:24:34][chrome #1]      takesHeapSnapshot: true,
[10:24:34][chrome #1]      takesScreenshot: true,
[10:24:34][chrome #1]      version: '39.0.2171.99',
[10:24:34][chrome #1]      webStorageEnabled: true } }
[10:24:34][chrome #1] 
[10:24:34]Running Jasmine Tests
[10:24:34]
[10:24:34][launcher] Runner process exited unexpectedly with error code: 1
[10:24:34][launcher] 1 instance(s) of WebDriver still running
[10:28:03]........................... [31m>>  [39m
[10:28:03] [31mFatal error: protractor exited with code: 100  [39m
```

Looks like it's still crashing at different points. I'm beginning to think it's something to do with my CI server given the flakiness.

**Comment by juliemr** [comment]
Can this be closed?

**Comment by JoshSchreuder** [comment]
@juliemr sure thing, I still don't know what's causing the issue, but I have a feeling it might be my CI server rather than Protractor, because I still can't reliably reproduce it outside of a build.

**Comment by mjaverto** [comment]
@JoshSchreuder did you make any progress on finding your issue? I'm having a similar issues using Atlassian's Bamboo and OS X. Seems to only happen with CI.

**Comment by JoshSchreuder** [comment]
@mjaverto unfortunately not, I'm still having these issues. As of Protractor 2.0.0 I'm getting:

```
[14:42:41][chrome #1] PID: 4572
[14:42:41][chrome #1] Using ChromeDriver directly...
[14:42:41][chrome #1] undefined is not a function
```

and

```
[14:43:39][firefox #2] PID: 1236
[14:43:39][firefox #2] Using FirefoxDriver directly...
[14:43:39][firefox #2] undefined is not a function
```

which is odd.

**Comment by mjaverto** [comment]
@JoshSchreuder I think we may have found the issue but it is over a month later so I'm not 100% sure this was the fix.. 

For OS X we logged in as the user who is running Bamboo. 

If that user wasn't logged in, they didn't have a session and couldn't open any windows. When that user is logged in (even if locked) it seems this works without issue. The reason I noticed was when we diagnosed one day, the user was logged in and chrome windows were opening as a bamboo job built in the background.
